docs/two-phase-commit-transactions.md

```diff
@@ -195,6 +195,111 @@ Similar to `prepare()`, you can call `validate()` in the coordinator/participant
+        try {
+          tx.rollback();
+        } catch (RollbackException ex) {
+          // Rolling back the transaction fails. You can log it here
+        }
+        return;
+      } catch (UnknownTransactionStatusException e) {
+        // If you catch `UnknownTransactionStatusException` when committing the transaction, you are
+        // not sure if the transaction succeeds or not. In such a case, you need to check if the
+        // transaction is committed successfully or not and retry it if it failed. How to identify a
+        // transaction status is delegated to users
+        return;
+      }
+    }
+  }
+}
+```
+
+The APIs for CRUD operations (`get()`/`scan()`/`put()`/`delete()`/`mutate()`) could throw `CrudException` and `CrudConflictException`.
+If you catch `CrudException`, it indicates some failure (e.g., database failure and network error) happens during a transaction, so you should cancel the transaction or retry the transaction after the failure/error is fixed.
+If you catch `CrudConflictException`, it indicates a conflict occurs during the transaction so that you can retry the transaction from the beginning, preferably with well-adjusted exponential backoff based on your application and environment.
+The sample code retries three times maximum and sleeps 100 milliseconds before retrying the transaction.
+
+The `prepare()` API could throw `PreparationException` and `PreparationConflictException`.
+If you catch `PreparationException`, like the `CrudException` case, you should cancel the transaction or retry the transaction after the failure/error is fixed.
+If you catch `PreparationConflictException`, like the `CrudConflictException` case, you can retry the transaction from the beginning.
+
+The `validate()` API could throw `ValidationException` and `ValidationConflictException`.
+If you catch `ValidationException`, like the `CrudException` case, you should cancel the transaction or retry the transaction after the failure/error is fixed.
+If you catch `ValidationConflictException`, like the `CrudConflictException` case, you can retry the transaction from the beginning.
+
+The `commit()` API could throw `CommitException`, `CommitConflictException`, and `UnknownTransactionStatusException`.
+If you catch `CommitException`, like the `CrudException` case, you should cancel the transaction or retry the transaction after the failure/error is fixed.
+If you catch `CommitConflictException`, like the `CrudConflictException` case, you can retry the transaction from the beginning.
+If you catch `UnknownTransactionStatusException`, you are not sure if the transaction succeeds or not.
+In such a case, you need to check if the transaction is committed successfully or not and retry it if it fails.
+How to identify a transaction status is delegated to users.
+You may want to create a transaction status table and update it transactionally with other application data so that you can get the status of a transaction from the status table.
```

Comment:
I don't really get this. Could you tell me a simple example?

Reply:
This is explained in the 3rd way in the following article (Japanese):
http://nippondanji.blogspot.com/2009/02/blog-post_23.html

You can have a status (or log) table and insert a record into this table in addition to the usual operations in a transaction. And when you get `UnknownTransactionStatusException` and want to check if the transaction is really committed, you can check if the record of the status table exists or not. If the record exists, the transaction has been successfully committed.

Reply:
Great! Thank you for sharing it.

The blog post seems really helpful for our customers as I think it's hard for them to come up with the idea without the details. How about adding a link to similar articles here? I with if the blog post were written in English...

Reply:
BTW (a bit off-topic), do you think it would be great if ScalarDB maintains this kind of transaction history in `coordinator.history` or something instead of asking application side? ScalarDB error handling can check the table when trying to throw UnknownTransactionStatusException.

Reply:
> coordinator.history

Ah, checking `coordinator.state` might be enough. But the point in this case is that maybe something wrong with the underlying database is happening and it's likely ScalarDB can't access the table... 🤔

Reply:
[off-topic]

> This is explained in the 3rd way in the following article (Japanese):
> http://nippondanji.blogspot.com/2009/02/blog-post_23.html

Just remembered I did the same way before.
https://github.com/treasure-data/digdag/blob/42ea7cb71270957ea6424c2df88dd7c65d248789/digdag-standards/src/main/java/io/digdag/standards/operator/redshift/RedshiftConnection.java#L297-L303 The workflow operator for Redshift does CRUD in the user's database and implicitly inserts a status log record to a secret table in the same transaction

Reply:
> Ah, checking coordinator.state might be enough. But the point in this case is that maybe something wrong with the underlying database is happening and it's likely ScalarDB can't access the table... 🤔

Yes, we have `coordinator.state` which has a state record for each transaction, and we can check the record to see if the transaction is committed or not. And we actually expose it through [DistributedTransactionManager.getState()](https://github.com/scalar-labs/scalardb/blob/18c8b20984c97b67b70b0a4f532c7e0ec23f698a/core/src/main/java/com/scalar/db/api/DistributedTransactionManager.java#L198).

However, I would like to deprecate (and remove) this API in the future because of the following reasons:

1. The `coordinator.state` table is an implementation-specific thing (the Consensus Commit transaction has it, but the JDBC transaction doesn't have it).
2. We don't want to keep the state records in the `coordinator.state` table permanently. After successfully committing or rolling back a transaction, we can essentially remove the state record of the transaction (except for the `getState()` reason). So in the future, I want to remove a state record after the transaction is complete.
3. We plan to implement the one-phase commit optimization. In this optimization, we don't need to write the state record to the `coordinator.state` table.

Reply:
I see. Regarding 1, I don't think we need to expose it to users as far as we can call it internally considering the mode. As for 2 and 3, it would matter. Also, as I mentioned above, automatically internal checking `state` (or `history`) table doesn't work when underlying database isn't working fine. So, it's difficult to suggest the best way to handle this kind of situation. I think this PR overall looks good. Probably we can get feedbacks from customers about this section in the future and let's discuss it deeply again.